txnsync/encodedgroupsmarshalers.go

```diff
@@ -0,0 +1,818 @@
+		}
+		stub.BitmaskSig.SetBit(i)
+		stub.Sig = append(stub.Sig, txn.Lsig.Sig[:]...)
+	}
+	if txn.Lsig.Msig.Version != 0 {
+		if len(stub.BitmaskVersion) == 0 {
+			stub.BitmaskVersion = make(bitmask, bitmaskLen)
+			stub.Version = make([]byte, 0, int(stub.TotalTransactionsCount))
+		}
+		stub.BitmaskVersion.SetBit(i)
+		stub.Version = append(stub.Version, txn.Lsig.Msig.Version)
+	}
+	if txn.Lsig.Msig.Threshold != 0 {
+		if len(stub.BitmaskThreshold) == 0 {
+			stub.BitmaskThreshold = make(bitmask, bitmaskLen)
+			stub.Threshold = make([]byte, 0, int(stub.TotalTransactionsCount))
+		}
+		stub.BitmaskThreshold.SetBit(i)
+		stub.Threshold = append(stub.Threshold, txn.Lsig.Msig.Threshold)
+	}
+	if txn.Lsig.Msig.Subsigs != nil {
+		if len(stub.BitmaskSubsigs) == 0 {
+			stub.BitmaskSubsigs = make(bitmask, bitmaskLen)
+			stub.Subsigs = make([][]crypto.MultisigSubsig, 0, int(stub.TotalTransactionsCount))
+		}
+		stub.BitmaskSubsigs.SetBit(i)
+		stub.Subsigs = append(stub.Subsigs, txn.Lsig.Msig.Subsigs)
+	}
+}
+
+func (stub *txGroupsEncodingStub) finishDeconstructLsigs() {
+	stub.BitmaskLogic.trimBitmask(int(stub.TotalTransactionsCount))
+	stub.BitmaskLogicArgs.trimBitmask(int(stub.TotalTransactionsCount))
+}
+
+func (stub *txGroupsEncodingStub) deconstructTransactions(i int, txn transactions.SignedTxn) error {
+	bitmaskLen := bytesNeededBitmask(int(stub.TotalTransactionsCount))
+	txTypeByte, err := TxTypeToByte(txn.Txn.Type)
+	if err != nil {
+		return fmt.Errorf("failed to deconstructTransactions: %w", err)
+	}
+	if len(stub.BitmaskTxType) == 0 {
+		stub.BitmaskTxType = make(bitmask, bitmaskLen)
+		stub.TxType = make([]byte, 0, int(stub.TotalTransactionsCount))
+	}
+	stub.TxType = append(stub.TxType, txTypeByte)
+	stub.deconstructTxnHeader(i, txn)
+	switch txn.Txn.Type {
+	case protocol.PaymentTx:
+		stub.deconstructPaymentTxnFields(i, txn)
+	case protocol.KeyRegistrationTx:
+		stub.deconstructKeyregTxnFields(i, txn)
+	case protocol.AssetConfigTx:
+		stub.deconstructAssetConfigTxnFields(i, txn)
+	case protocol.AssetTransferTx:
+		stub.deconstructAssetTransferTxnFields(i, txn)
+	case protocol.AssetFreezeTx:
+		stub.deconstructAssetFreezeTxnFields(i, txn)
+	case protocol.ApplicationCallTx:
+		stub.deconstructApplicationCallTxnFields(i, txn)
+	case protocol.CompactCertTx:
+		stub.deconstructCompactCertTxnFields(i, txn)
+	}
+	return nil
+}
+
+func (stub *txGroupsEncodingStub) finishDeconstructTransactions() {
+	stub.finishDeconstructTxType()
+	stub.finishDeconstructTxnHeader()
+	stub.finishDeconstructKeyregTxnFields()
+	stub.finishDeconstructPaymentTxnFields()
+	stub.finishDeconstructAssetConfigTxnFields()
+	stub.finishDeconstructAssetTransferTxnFields()
+	stub.finishDeconstructAssetFreezeTxnFields()
+	stub.finishDeconstructApplicationCallTxnFields()
+	stub.finishDeconstructCompactCertTxnFields()
+}
+
+func (stub *txGroupsEncodingStub) finishDeconstructTxType() {
+	offset := byte(0)
+	count := make(map[int]int)
```

Comment:
given that there are exactly 8 transaction types, you'd do *much* better allocating a `[8]int` instead of using a map here.